### PR TITLE
SDL 0118  Video Streaming Backgrounded String

### LINKED ISSUE #1058: [SDL 0118] Video Streaming Backgrounded String
Proposal: [Video Streaming Backgrounded String](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0118-video-background-string.md)

> Display a string of text on the head unit screen when the app leaves the foreground.

Review: https://github.com/smartdevicelink/sdl_evolution/issues/356

Steering Committee Decision:

> The Steering Committee voted to accept this proposal with revisions. The revisions will include updating the text to reflect Ford legal's preference, as described in this comment from Ford.  There was also discussion on the call regarding translations of this text. It was determined these could be discussed at a later date upon implementation, but would likely not require a separate Evolution Proposal for review.
> The proposal .md file was updated to reflect these revisions on 8/27/18.

### PR DESCRIPTION
**Pull request by NicoleYarroch**
Fixes #1058 

This PR is **ready** for review.

### Risk
This PR makes **minor** API changes.

### Testing Plan
Test cases added to the _SDLStreamingVideoLifecycleManagerSpec_

### Summary
The `SDLStreamingVideoLifecycleManager` now displays a string of text on the head unit screen when the app leaves the foreground.

##### Enhancements
* When a video streaming `NAVIGATION` or `PROJECTION` app enters the background on the phone, the video stream stops sending data due to background limitations. On iOS devices as the device is entering the background, a few frames of pure black are sent. Now we send a black screen with some white text telling the driver why their stream stopped.

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

**Comment by joeljfischer** [comment]
@NicoleYarroch Please resolve the merge conflict

**Comment by NicoleYarroch** [comment]
> @NicoleYarroch Please resolve the merge conflict

Fixed.

File: SmartDeviceLink/SDLStreamingMediaManager.h [hunk 1]
```diff
@@ -118,7 +119,17 @@ NS_ASSUME_NONNULL_BEGIN
...
+
+/**
+ Create a new streaming media manager for navigation and VPM apps with a specified configuration
+
+ @param connectionManager The pass-through for RPCs
+ @param streamingMediaConfiguration The configuration of this streaming media session
+ @param lifecycleConfiguration The lifecycle configuration of the app
+ @return A new streaming manager
+ */
+- (instancetype)initWithConnectionManager:(id<SDLConnectionManagerType>)connectionManager streamingMediaConfiguration:(SDLStreamingMediaConfiguration *)streamingMediaConfiguration lifecycleConfiguration:(SDLLifecycleConfiguration *)lifecycleConfiguration NS_DESIGNATED_INITIALIZER;
```
Comment (joeljfischer) [inline SmartDeviceLink/SDLStreamingMediaManager.h]: What about just passing the `SDLConfiguration`?

File: SmartDeviceLink/SDLStreamingVideoLifecycleManager.h [hunk 2]
```diff
@@ -138,7 +139,17 @@ NS_ASSUME_NONNULL_BEGIN
...
+
+/**
+ Create a new streaming media manager for navigation and VPM apps with a specified configuration
+
+ @param connectionManager The pass-through for RPCs
+ @param streamingMediaConfiguration The configuration of this streaming media session
+ @param lifecycleConfiguration The lifecycle configuration of the app
+ @return A new streaming manager
+ */
+- (instancetype)initWithConnectionManager:(id<SDLConnectionManagerType>)connectionManager streamingMediaConfiguration:(SDLStreamingMediaConfiguration *)streamingMediaConfiguration lifecycleConfiguration:(SDLLifecycleConfiguration *)lifecycleConfiguration NS_DESIGNATED_INITIALIZER;
```
Comment (joeljfischer) [inline SmartDeviceLink/SDLStreamingVideoLifecycleManager.h]: See above

File: SmartDeviceLink/SDLStreamingVideoLifecycleManager.m [hunk 3]
```diff
@@ -774,6 +781,10 @@ - (BOOL)isHmiStateVideoStreamCapable {
     return @[h264raw, h264rtp];
 }
 
+- (NSString *)videoStreamBackgroundString {
```
Comment (joeljfischer) [inline SmartDeviceLink/SDLStreamingVideoLifecycleManager.m]: This should probably be a getter only property.

File: SmartDeviceLink/SDLStreamingVideoLifecycleManager.m [hunk 4]
```diff
@@ -774,6 +781,10 @@ - (BOOL)isHmiStateVideoStreamCapable {
...
+    return [NSString stringWithFormat:@"When it is safe to do so, open %@ on phone", self.appName];
```
Comment (joeljfischer) [inline SmartDeviceLink/SDLStreamingVideoLifecycleManager.m]: Please add "your" before "phone"

File: SmartDeviceLink/SDLStreamingMediaManager.h (hunk 1 above)
Comment (NicoleYarroch) [reply SmartDeviceLink/SDLStreamingMediaManager.h]: Fixed

File: SmartDeviceLink/SDLStreamingVideoLifecycleManager.h (hunk 2 above)
Comment (NicoleYarroch) [reply SmartDeviceLink/SDLStreamingVideoLifecycleManager.h]: Fixed

File: SmartDeviceLink/SDLStreamingVideoLifecycleManager.m (hunk 3 above)
Comment (NicoleYarroch) [reply SmartDeviceLink/SDLStreamingVideoLifecycleManager.m]: Fixed

File: SmartDeviceLink/SDLStreamingVideoLifecycleManager.m (hunk 4 above)
Comment (NicoleYarroch) [reply SmartDeviceLink/SDLStreamingVideoLifecycleManager.m]: Fixed

File: SmartDeviceLink/SDLStreamingVideoLifecycleManager.m [hunk 5]
```diff
@@ -66,45 +71,54 @@ @interface SDLStreamingVideoLifecycleManager() <SDLVideoEncoderDelegate>
...
 
+@property (copy, nonatomic) NSString *appName;
 @property (assign, nonatomic) CV_NULLABLE CVPixelBufferRef backgroundingPixelBuffer;
 
 @property (strong, nonatomic, nullable) CADisplayLink *displayLink;
 @property (assign, nonatomic) BOOL useDisplayLink;
 
 @property (assign, nonatomic) CMTime lastPresentationTimestamp;
 
+@property (copy, nonatomic, getter=getVideoStreamBackgroundString) NSString *videoStreamBackgroundString;
```
Comment (joeljfischer) [inline SmartDeviceLink/SDLStreamingVideoLifecycleManager.m]: Objective-C getters best practice does not use `get` nomenclature. Remove the `getter=getVideoStreamBackgroundString` and reference the method with just the `videoStreamBackgroundString `.

File: SmartDeviceLink/SDLStreamingVideoLifecycleManager.h [hunk 6]
```diff
@@ -138,7 +139,16 @@ NS_ASSUME_NONNULL_BEGIN
  @param configuration The configuration of this streaming media session
  @return A new streaming manager
  */
-- (instancetype)initWithConnectionManager:(id<SDLConnectionManagerType>)connectionManager configuration:(SDLStreamingMediaConfiguration *)configuration NS_DESIGNATED_INITIALIZER;
+- (instancetype)initWithConnectionManager:(id<SDLConnectionManagerType>)connectionManager configuration:(SDLStreamingMediaConfiguration *)configuration __deprecated_msg("Use initWithConnectionManager:streamingMediaConfiguration:lifecycleConfiguration: instead");
```
Comment (joeljfischer) [inline SmartDeviceLink/SDLStreamingVideoLifecycleManager.h]: This file / class is private, so it's not necessary to deprecate a method here. You can just update this method and remove the old one.

File: SmartDeviceLinkTests/DevAPISpecs/SDLStreamingVideoLifecycleManagerSpec.m [hunk 7]
```diff
@@ -32,15 +33,21 @@
 #import "SDLVideoStreamingState.h"
 #import "TestConnectionManager.h"
 
+@interface SDLStreamingVideoLifecycleManager ()
+@property (copy, nonatomic) NSString *appName;
+- (NSString *)videoStreamBackgroundString;
```
Comment (joeljfischer) [inline SmartDeviceLinkTests/DevAPISpecs/SDLStreamingVideoLifecycleManagerSpec.m]: This needs to update along with the change to this method (by turning it into a property) above

Comment (NicoleYarroch) [reply SmartDeviceLinkTests/DevAPISpecs/SDLStreamingVideoLifecycleManagerSpec.m]: Fixed

File: SmartDeviceLink/SDLStreamingVideoLifecycleManager.h (hunk 6 above)
Comment (NicoleYarroch) [reply SmartDeviceLink/SDLStreamingVideoLifecycleManager.h]: Fixed

File: SmartDeviceLink/SDLStreamingVideoLifecycleManager.m (hunk 5 above)
Comment (NicoleYarroch) [reply SmartDeviceLink/SDLStreamingVideoLifecycleManager.m]: Fixed

File: SmartDeviceLink/SDLStreamingVideoLifecycleManager.m [hunk 8]
```diff
@@ -66,45 +68,49 @@ @interface SDLStreamingVideoLifecycleManager() <SDLVideoEncoderDelegate>
...
 
+@property (copy, nonatomic) NSString *appName;
 @property (assign, nonatomic) CV_NULLABLE CVPixelBufferRef backgroundingPixelBuffer;
 
 @property (strong, nonatomic, nullable) CADisplayLink *displayLink;
 @property (assign, nonatomic) BOOL useDisplayLink;
 
 @property (assign, nonatomic) CMTime lastPresentationTimestamp;
 
+@property (copy, nonatomic) NSString *videoStreamBackgroundString;
```
Comment (joeljfischer) [inline SmartDeviceLink/SDLStreamingVideoLifecycleManager.m]: This should be `readonly`.

Comment (NicoleYarroch) [reply SmartDeviceLink/SDLStreamingVideoLifecycleManager.m]: Fixed

**Pull request merged**